### PR TITLE
add `--keep-debug-symbols` configuration option to set default value of '`debug`' toolchain option, and enable it by default so `-g` is included in `$CXXFLAGS` & co

### PR DESCRIPTION
**Pull request by Micket**
Also defaults to "true" as was discussed for 5.0.x. 
Those concerned with disk space can consider switching this back to false. 

Disclaimer: It's getting late and I'm really not sure if i can get away with inserting the default value of "debug" in the compiler option map like that. Not sure if build_option is guaranteed to be initialized when that static dict is defined.

### CODE REVIEW

**Comment by Micket** [comment]
OK the optparser thing won't permit setting bools in any way. Listing options they get converted to strings, which is ugly to work with, and optparse doesn't support `bool` types for selection. So it's a enabling flag instead.

Also, due to how classes are sneakily initialized by EB, the static class variables aren't always defined, so, they might be `None`. As existing codes just `if self.xxxx is not None:`  then I guess i'll just do the same.

File: easybuild/tools/options.py [hunk 1]
```diff
@@ -392,6 +392,7 @@ def override_options(self):
             'debug-lmod': ("Run Lmod modules tool commands in debug module", None, 'store_true', False),
             'default-opt-level': ("Specify default optimisation level", 'choice', 'store', DEFAULT_OPT_LEVEL,
                                   Compiler.COMPILER_OPT_FLAGS),
+            'default-debug-symbols': ("Sets default value of debug toolchain option", None, 'store_true', True),
```
Comment (boegel) [inline easybuild/tools/options.py:395]: Maybe `retain-debug-symbols-by-default` is a better name for this configuration option?

Enabling it by default makes sense I think, there *should* be no fallout of that...

File: easybuild/tools/toolchain/compiler.py [hunk 2]
```diff
@@ -82,7 +82,7 @@ class Compiler(Toolchain):
         'loose': (False, "Loose precision"),
         'veryloose': (False, "Very loose precision"),
         'verbose': (False, "Verbose output"),
-        'debug': (False, "Enable debug"),
+        'debug': (None, "Enable or disables debug symbols"),  # refers to default-debug-symbols build option by default
```
Comment (boegel) [inline easybuild/tools/toolchain/compiler.py:85]:
```suggestion
        # default value for 'debug' toolchain option is determined by retain-debug-symbols-by-default configuration option
        'debug': (None, "Retain debug symbols"),
```

File: easybuild/tools/toolchain/compiler.py [hunk 3]
```diff
@@ -175,6 +175,10 @@ def set_variables(self):
 
     def _set_compiler_toolchainoptions(self):
         """Set the compiler related toolchain options"""
+        # Initialize default value of debug symbols based on global build option
```
Comment (boegel) [inline easybuild/tools/toolchain/compiler.py:178]:
```suggestion
        # Initialize default value of 'debug' toolchain option based on global configuration option option
```

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/tools/options.py (hunk 1 above)
Comment (Micket) [reply easybuild/tools/options.py:395]: Well I won't insist, though I find the wording of "retain" a bit strange, reads to me that we were intentionally stripping them out.

Comment (boegel) [reply easybuild/tools/options.py:395]: As agreed in sync call:
```suggestion
            'keep-debug-symbols': ("Sets default value of debug toolchain option", None, 'store_true', True),
```

**Review by boegel — CHANGES_REQUESTED** [review]

**Review by boegel — APPROVED** [review]

**Pull request merged**